sync/http/route.go

```diff
@@ -1,130 +1,197 @@
-	return NewRoute(p, http.MethodGet, pr, trace, nil, mm...)
+// RouteBuilder for building a route.
+type RouteBuilder struct {
+	method        string
+	path          string
+	trace         bool
+	middlewares   []MiddlewareFunc
+	authenticator auth.Authenticator
+	handler       http.HandlerFunc
+	errors        []error
 }
 
-// NewPostRoute creates a new POST route from a generic handler.
-func NewPostRoute(p string, pr sync.ProcessorFunc, trace bool, mm ...MiddlewareFunc) Route {
-	return NewRoute(p, http.MethodPost, pr, trace, nil, mm...)
+// WithTrace enables route tracing.
+func (rb *RouteBuilder) WithTrace() *RouteBuilder {
+	rb.trace = true
+	return rb
 }
 
-// NewPutRoute creates a new PUT route from a generic handler.
-func NewPutRoute(p string, pr sync.ProcessorFunc, trace bool, mm ...MiddlewareFunc) Route {
-	return NewRoute(p, http.MethodPut, pr, trace, nil, mm...)
+// WithMiddlewares adds middlewares.
+func (rb *RouteBuilder) WithMiddlewares(mm ...MiddlewareFunc) *RouteBuilder {
+	if len(mm) == 0 {
+		rb.errors = append(rb.errors, errors.New("middlewares are empty"))
+	}
+	rb.middlewares = mm
+	return rb
 }
 
-// NewDeleteRoute creates a new DELETE route from a generic handler.
-func NewDeleteRoute(p string, pr sync.ProcessorFunc, trace bool, mm ...MiddlewareFunc) Route {
-	return NewRoute(p, http.MethodDelete, pr, trace, nil, mm...)
+// WithAuth adds authenticator.
+func (rb *RouteBuilder) WithAuth(auth auth.Authenticator) *RouteBuilder {
+	if auth == nil {
+		rb.errors = append(rb.errors, errors.New("authenticator is nil"))
+	}
+	rb.authenticator = auth
+	return rb
 }
 
-// NewPatchRoute creates a new PATCH route from a generic handler.
-func NewPatchRoute(p string, pr sync.ProcessorFunc, trace bool, mm ...MiddlewareFunc) Route {
-	return NewRoute(p, http.MethodPatch, pr, trace, nil, mm...)
+// WithMethodGet HTTP method.
+func (rb *RouteBuilder) WithMethodGet() *RouteBuilder {
+	rb.method = http.MethodGet
+	return rb
 }
 
-// NewHeadRoute creates a new HEAD route from a generic handler.
-func NewHeadRoute(p string, pr sync.ProcessorFunc, trace bool, mm ...MiddlewareFunc) Route {
-	return NewRoute(p, http.MethodHead, pr, trace, nil, mm...)
+// WithMethodHead HTTP method.
+func (rb *RouteBuilder) WithMethodHead() *RouteBuilder {
+	rb.method = http.MethodHead
+	return rb
 }
 
-// NewOptionsRoute creates a new OPTIONS route from a generic handler.
-func NewOptionsRoute(p string, pr sync.ProcessorFunc, trace bool, mm ...MiddlewareFunc) Route {
-	return NewRoute(p, http.MethodOptions, pr, trace, nil, mm...)
+// WithMethodPost HTTP method.
+func (rb *RouteBuilder) WithMethodPost() *RouteBuilder {
+	rb.method = http.MethodPost
+	return rb
 }
 
-// NewRoute creates a new route from a generic handler with auth capability.
-func NewRoute(p string, m string, pr sync.ProcessorFunc, trace bool, auth auth.Authenticator, mm ...MiddlewareFunc) Route {
+// WithMethodPut HTTP method.
```

Comment:
Given that now we can call these methods sequentially, but we can only have one method, would it make sense to check if rb.method is set , and potentially throw an error ? We might also consider changing the name of these methods to remove the "WithMethod" part. What do you think ?

Reply:
About the error i agree!
What is your proposal for the naming?

Reply:
maybe just 'Put' , 'Get' etc ... ?

Reply:
I did rename from `WithMethodGet()` to `MethodGet()`. Using just `Get()` looked like a getter method and was not clear as an API.